cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp

```diff
@@ -0,0 +1,977 @@
+        localBox[1]         = -daal::services::internal::MaxVal<DataType>::get();
+        localBox[2]         = daal::services::internal::MaxVal<DataType>::get();
+        localBox[3]         = -daal::services::internal::MaxVal<DataType>::get();
+        return localBox;
+    });
+    const IdxType nThreads    = tlsBox.nthreads();
+    const IdxType sizeOfBlock = services::internal::min<cpu, IdxType>(256, N / nThreads + 1);
+    const IdxType nBlocks     = N / sizeOfBlock + !!(N % sizeOfBlock);
+
+    daal::static_threader_for(nBlocks, [&](IdxType iBlock, IdxType tid) {
+        const IdxType iStart = iBlock * sizeOfBlock;
+        const IdxType iEnd   = services::internal::min<cpu, IdxType>(N, iStart + sizeOfBlock);
+        DataType * localBox  = tlsBox.local(tid);
+
+        for (IdxType i = iStart; i < iEnd; ++i)
+        {
+            localBox[0] = services::internal::min<cpu, DataType>(localBox[0], posx[i]);
+            localBox[1] = services::internal::max<cpu, DataType>(localBox[1], posx[i]);
+            localBox[2] = services::internal::min<cpu, DataType>(localBox[2], posy[i]);
+            localBox[3] = services::internal::max<cpu, DataType>(localBox[3], posy[i]);
+        }
+    });
+
+    tlsBox.reduce([&](DataType * ptr) -> void {
+        if (!ptr) return;
+        box[0] = services::internal::min<cpu, DataType>(box[0], ptr[0]);
+        box[1] = services::internal::max<cpu, DataType>(box[1], ptr[1]);
+        box[2] = services::internal::min<cpu, DataType>(box[2], ptr[2]);
+        box[3] = services::internal::max<cpu, DataType>(box[3], ptr[3]);
+        services::internal::service_free<DataType, cpu>(ptr);
+    });
+
+    //scale the maximum to get all points strictly in the bounding box
+    if (box[1] >= 0.)
+        box[1] = (box[1] * (1. + 1e-3));
+    else
+        box[1] = (box[1] * (1. - 1e-3));
+    if (box[3] >= 0.)
+        box[3] = (box[3] * (1. + 1e-3));
+    else
+        box[3] = (box[3] * (1. - 1e-3));
+
+    //save results
+    radius       = services::internal::max<cpu, DataType>(box[1] - box[0], box[3] - box[2]) * 0.5;
+    posx[nNodes] = (box[0] + box[1]) * 0.5;
+    posy[nNodes] = (box[2] + box[3]) * 0.5;
+
+    return services::Status();
+}
+
+template <typename IdxType, typename DataType, daal::CpuType cpu>
+services::Status qTreeBuildingKernelImpl(IdxType * child, const DataType * posx, const DataType * posy, IdxType * duplicates, const IdxType nNodes,
+                                         const IdxType N, IdxType & maxDepth, IdxType & bottom, const DataType & radius)
+{
+    // internal variables
+    IdxType j      = 0;
+    IdxType depth  = 0;
+    IdxType ch     = 0;
+    IdxType n      = 0;
+    IdxType locked = 0;
+    IdxType patch  = 0;
+    DataType x     = 0.;
+    DataType y     = 0.;
+    DataType r     = 0.;
+    DataType px    = 0.;
+    DataType py    = 0.;
+
+    //initialize array
+    services::internal::service_memset<IdxType, cpu>(child, -1, (nNodes + 1) * 4);
```

Comment:
In places where might be overflow (most often this is multiplication/addition) additional check/assert is needed. You may only need to check once.